### PR TITLE
Add tests to ensure CAPE.depositErc20 and CAPE.submitBlock are non reentrant.

### LINKED ISSUE #790: Test correct use of ReentrancyGuard
Create a malicious token that makes a re-entrant call with an empty block. Check revert reason.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/52eeebecda140ebaf4ec8752ed119d8288287fac/contracts/security/ReentrancyGuard.sol#L52

### PR DESCRIPTION
**Pull request by philippecamacho**
Closes #790.

### CODE REVIEW

File: contracts/deploy/10_token.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
   const { deployer } = await getNamedAccounts();
 
   await deploy("SimpleToken", { from: deployer, log: true });
+  await deploy("MaliciousToken", { from: deployer, log: true });
```
Comment (sveitser) [inline contracts/deploy/10_token.ts:10]: Is it useful to deploy the malicious token with hardhat? If yes, would it be better to put it into it's own deployment file so we can still do a deployed with `SimpleToken` but without `MaliciousToken` if needed?

File: contracts/contracts/MaliciousToken.sol [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     * /// @dev Sets the address for performing the reentrancy attack.
+     */
+    function setTargetContractAddress(address targetContractAddress) public {
+        _targetContractAddress = targetContractAddress;
+    }
+
+    /**
+     * /// @dev pick the depositErc20 function when calling back the CAPE contract
+     */
+    function runDeposit() public {
```
Comment (sveitser) [inline contracts/contracts/MaliciousToken.sol:39]:
```suggestion
    function enableDepositAttack() public {
```
Same suggestion for `runSubmitBlock`.

With the current naming I thought this runs the deposit function when looking at the test.

File: contracts/rust/src/cape/reentrancy.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        cape_contract.address(),
+        Arc::new(owner_of_erc20_tokens_client.clone()),
+    );
+
+    let sponsor = owner_of_erc20_tokens_client_address;
+
+    let description = erc20_asset_description(&erc20_code, &EthereumAddr(sponsor.to_fixed_bytes()));
+    let asset_code = AssetCode::new_foreign(&description);
+    let asset_def = AssetDefinition::new(asset_code, AssetPolicy::rand_for_test(rng)).unwrap();
+    let asset_def_sol = asset_def.clone().generic_into::<sol::AssetDefinition>();
```
Comment (sveitser) [inline contracts/rust/src/cape/reentrancy.rs:44]: I think a few variables here aren't so useful: `asset_def_sol`, `amount_u256`,  `cape_contract_address`, `owner_of_erc20_tokens_client_address`, `asset_code`, `call`.

**Review by sveitser — APPROVED** [review]
LGTM. Just some comments. Nice test to have :+1:

File: contracts/deploy/10_token.ts (hunk 1 above)
Comment (philippecamacho) [reply contracts/deploy/10_token.ts]: Yes, a new script for the Malicious Token has been added in 51cd9d0bc1b1accc21ce6f172e09fae451bee9c4.

File: contracts/rust/src/cape/reentrancy.rs (hunk 3 above)
Comment (philippecamacho) [reply contracts/rust/src/cape/reentrancy.rs]: These variables have been removed in faa84b3c812dda1a134ba3c7d15d5ee4e15ba23b.

**Pull request merged**